### PR TITLE
feat(event_commands.go): add clone command

### PR DESCRIPTION
**Pull request by vdice**
**What this PR does / why we need it**:

- Adds the `brig event clone` command per https://github.com/brigadecore/brigade/issues/1196

Closes https://github.com/brigadecore/brigade/issues/1196

**Special notes for your reviewer**:

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

### CODE REVIEW

**Comment by krancour** [comment]
I'm wondering if this is another case wherein we should move the clone logic to the backend rather than rely on all potential clients to get it right.

**Comment by vdice** [comment]
> I'm wondering if this is another case wherein we should move the clone logic to the backend rather than rely on all potential clients to get it right.

Sure.  We'd probably want to do the same for retry as well, huh?  I can add a note to that ticket as well.

**Comment by krancour** [comment]
> Sure. We'd probably want to do the same for retry as well, huh? I can add a note to that ticket, if so.

I believe so.

**Comment by vdice** [comment]
Ok, now updated with clone logic handled by the backend event service.  PTAL when convenient @krancour

File: sdk/v2/core/events.go [hunk 1]
```diff
@@ -232,6 +232,13 @@ type EventsClient interface {
 	List(context.Context, *EventsSelector, *meta.ListOptions) (EventList, error)
 	// Get retrieves a single Event specified by its identifier.
 	Get(context.Context, string) (Event, error)
+	// Clones a pre-existing Event, removing the original's metadata and Worker
+	// config in the process. If the original Event references a Project by ID,
+	// one new Event will be created. Otherwise, the Event provided is treated
+	// as a template and zero or more discrete Events may be created-- one for
+	// each subscribed Project. An EventList is returned containing all newly
+	// created Events.
```
Comment (krancour) [inline sdk/v2/core/events.go:240]: By the time an event has an ID, it's a discrete event, so copy(eventID) can't possibly be copying a template. I think everything is good and just the comment here has to change.

File: sdk/v2/core/events.go [hunk 2]
```diff
@@ -324,6 +331,22 @@ func (e *eventsClient) Get(
...
+func (e *eventsClient) Clone(
+	ctx context.Context,
+	id string,
+) (EventList, error) {
+	events := EventList{}
+	return events, e.ExecuteRequest(
+		ctx,
+		rm.OutboundRequest{
+			Method:      http.MethodPut,
+			Path:        fmt.Sprintf("v2/events/%s/clone", id),
```
Comment (krancour) [inline sdk/v2/core/events.go:343]: REST nit. Post to `.../clones` might be better. Given that an event could be cloned multiple times, and therefor _have_ multiple clones, "clones" is a collection of sorts. Adding something to a collection is, more conventionally, a post.

File: v2/apiserver/internal/core/events.go [hunk 3]
```diff
@@ -532,6 +535,29 @@ func (e *eventsService) GetByWorkerToken(
 	return event, nil
 }
 
+func (e *eventsService) Clone(
+	ctx context.Context,
+	id string,
+) (EventList, error) {
+	events := EventList{}
+
+	if err := e.authorize(ctx, system.RoleReader(), ""); err != nil {
```
Comment (krancour) [inline v2/apiserver/internal/core/events.go:544]: I think this is the wrong role. One would need to have `core.ProjectUser` (for the event's project) in order to be authorized to create any event at all.

File: v2/apiserver/internal/core/events.go [hunk 4]
```diff
@@ -532,6 +535,29 @@ func (e *eventsService) GetByWorkerToken(
...
+	if err != nil {
+		return events, errors.Wrapf(err, "error retrieving event %q from store", id)
+	}
+
+	// Clone all event details *except* metadata and worker config
+	clone := event
+	clone.ObjectMeta = meta.ObjectMeta{}
+	clone.Worker = Worker{}
+
+	return e.Create(ctx, clone)
```
Comment (krancour) [inline v2/apiserver/internal/core/events.go:558]: Nice re-use.

File: v2/apiserver/internal/core/rest/events_endpoints.go [hunk 5]
```diff
@@ -40,6 +40,12 @@ func (e *EventsEndpoints) Register(router *mux.Router) {
 		e.AuthFilter.Decorate(e.get),
 	).Methods(http.MethodGet)
 
+	// Clone event
+	router.HandleFunc(
+		"/v2/events/{id}/clone",
```
Comment (krancour) [inline v2/apiserver/internal/core/rest/events_endpoints.go:45]: See https://github.com/brigadecore/brigade/pull/1362/files#r616007108

**Comment by krancour** [comment]
Overall looks great. There are just a few minor things to fix.

One additional thought I had is I think it might be useful to add a `cloneOf` field (or similar name) to the `Event` struct (both server-side and in the SDK) so that there's some kind of traceability back to the original event.

File: v2/apiserver/internal/core/events.go (hunk 3 above)
Comment (vdice) [reply v2/apiserver/internal/core/events.go]: Here I decided to remove the authz prior to the `store.Get` call -- and just defer to the authz checks in the subsequent `Create()` call.  I noticed that `UpdateSourceState`, `Cancel`, `Delete` all follow this pattern.  WDYT?

**Comment by vdice** [comment]
> One additional thought I had is I think it might be useful to add a cloneOf field (or similar name) to the Event struct (both server-side and in the SDK) so that there's some kind of traceability back to the original event.

Oh that's an interesting idea.  I forgot to add this in the last commit.  It sounds like this addition has the potential to be useful, but I'm wary of adding until we have strong motivations to do so.  That being said, what are your thoughts of adding a label to the `Labels` field on a cloned `Event`, e.g. `cloneOf=<event id>` instead of adding a new field to the `Event` struct?

File: v2/apiserver/internal/core/events.go (hunk 3 above)
Comment (krancour) [reply v2/apiserver/internal/core/events.go]: Maybe just leave a comment that makes note of that?

**Review by krancour — APPROVED** [review]
LGTM

**Pull request merged**